### DISCUSSION TITLE
Why is there a TextMate grammar generated if LSP is supported?

### DISCUSSION THREAD

**Discussion by kaby76** [eclipse-langium/langium#716] [Q&A] [answered]
I've implemented several LSP servers. In all of them, I never define a TextMate file. As of LSP 3.16, semantic tokens supports the colorization of text, and it's fast. But, when I go to create the "hello-world" from template (see https://langium.org/docs/getting-started/ for instructions), a TM file is generated (`Writing textmate grammar to C:\msys64\home\Kenne\langium-gs\hello-world\syntaxes\hello-world.tmLanguage.json`). Does Langium support LSP?

**Comment by pluralia** [comment, marked as answer]
Hi @kaby76,

> Does Langium support LSP?

Yes, Langium supports LSP and semantic tokens in particular. Langium's grammar DSL is written in Langium (and uses semantic tokens), so you can see the Langium source code for an example use case:

1. Implement your custom [`SemanticTokensProvider`](https://github.com/langium/langium/blob/main/packages/langium/src/lsp/semantic-token-provider.ts#L65-L69) (for that it's enough to extend `AbstractSemanticTokenProvider` like Langium does it [here](https://github.com/langium/langium/blob/main/packages/langium/src/grammar/lsp/grammar-semantic-tokens.ts)) 
2. Register your custom service  in `<your-dsl-name>-module.ts` (like [here](https://github.com/langium/langium/blob/main/packages/langium/src/grammar/langium-grammar-module.ts#L38))

---

As for the TextMate file: this is not a replacement for semantic tokens, but rather an extra way to have highlighting. We generate also a Monarch file (for instance, a [Monarch file](https://github.com/langium/langium/blob/main/examples/statemachine/syntaxes/statemachine.monarch.ts) for the statemachine example, generation of which can be enabled in `langium-config.json` like [here](https://github.com/langium/langium/blob/main/examples/statemachine/langium-config.json#L12-L14)), that is one more way to have highlighting in Langium. You can choose whichever way you prefer or is more suitable for your purposes.

**Comment by msujew** [comment]
In addition to what @pluralia already said, the TextMate grammar has a bit of an advantage, as it doesn't require a call to the language server in order to display the correct text styling. It can apply the token style instantly, while the semantic token styling will always have a short delay. For vscode, the optimal solution is usually to mix TextMate generated tokens and semantic tokens from the LSP for token styling. For example, Microsoft does the [same for TypeScript](https://github.com/microsoft/vscode/tree/main/extensions/typescript-basics/syntaxes).
